helix-rest/src/main/java/org/apache/helix/rest/server/service/InstanceServiceImpl.java

```diff
@@ -222,10 +223,15 @@ protected PartitionHealth generatePartitionHealthMapFromZK() {
     // Only checks the instances are online with valid reports
     List<String> liveInstances =
         _dataAccessor.getChildNames(_dataAccessor.keyBuilder().liveInstances());
-    for (String instance : liveInstances) {
-      ZNRecord customizedHealth = _dataAccessor
-          .getProperty(_dataAccessor.keyBuilder().healthReport(instance, PARTITION_HEALTH_KEY))
-          .getRecord();
+    // Make a parallel batch call for getting all healthreports from ZK.
+    List<HelixProperty> healthReports = _dataAccessor.getProperty(liveInstances.stream()
+        .map(instance -> _dataAccessor.keyBuilder().healthReport(instance, PARTITION_HEALTH_KEY))
+        .collect(Collectors.toList()));
+    for (int i = 0; i < liveInstances.size(); i++) {
```

Comment:
First, you shouldn't make any assumption of the order of `List<HelixProperty>`, the for loop now assumes for each instance i you can get its cutomizedHealth at i index. 
Second, it's still a sequential call even using stream

Reply:
I think you are still not clear why I make this change:
1. The order is guaranteed in getProperty. Please take a look of our implementation. If the ZNRecord is able to read back, it will be null. This is also be a change may related to your work.
2. The for loop is just in memory operation, we just combined the N calls to 1 call send to ZK to reduce the traffic.

Reply:
I oversee the getProperty(List..) at first.  What if the result is null then your implementation will have NPE.

Reply:
Our implementations guarantees the list is not null. Otherwise, lots of our implementation in controller will have problem. 

If you feel the check is must have, we can add checks here.